File: script/make/protocol/lua/TitleProtocol.lua
TitleProtocol = {}

function TitleProtocol.encode(offset, protocol, data)
    if protocol == 11901 then
        local offset = offset
        local table = {}
        return table
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

function TitleProtocol.decode(offset, protocol, data)
    if protocol == 11901 then
        local offset = offset
        -- 称号列表
        local list = {}
        local listLength = string.unpack(">I2", data, offset)
        offset = offset + 2
        for listIndex = 1, listLength do
            -- 称号ID
            local titleId = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- 过期时间
            local expireTime = string.unpack(">I4", data, offset)
            offset = offset + 4
            list[listIndex] = {titleId = titleId, expireTime = expireTime}
        end
        return {list = list}
    elseif protocol == 11902 then
        local offset = offset
        -- 称号ID列表
        local list = {}
        local listLength = string.unpack(">I2", data, offset)
        offset = offset + 2
        for listIndex = 1, listLength do
            -- 称号ID
            local titleId = string.unpack(">I4", data, offset)
            offset = offset + 4
            list[listIndex] = {titleId = titleId}
        end
        return {list = list}
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

File: script/make/protocol/cs/TitleProtocol.cs
public static class TitleProtocol
{
    public static void Encode(System.Text.Encoding encoding, System.IO.BinaryWriter writer, System.UInt16 protocol, System.Collections.Generic.Dictionary<System.String, System.Object> data) 
    {
        switch (protocol) 
        {
            case 11901:
            {
                return;
            }
            default:throw new System.ArgumentException(System.String.Format("unknown protocol define: {0}", protocol));
        }
    }

    public static System.Collections.Generic.Dictionary<System.String, System.Object> Decode(System.Text.Encoding encoding, System.IO.BinaryReader reader, System.UInt16 protocol) 
    {
        switch (protocol) 
        {
            case 11901:
            {
                // 称号列表
                var listLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var list = new System.Collections.Generic.List<System.Object>(listLength);
                while (listLength-- > 0)
                {
                    // Title
                    // 称号ID
                    var titleId = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                    // 过期时间
                    var expireTime = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                    // object
                    var title = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"titleId", titleId}, {"expireTime", expireTime}};
                    // add
                    list.Add(title);
                }
                return new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"list", list}};
            }
            case 11902:
            {
                // 称号ID列表
                var listLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var list = new System.Collections.Generic.List<System.Object>(listLength);
                while (listLength-- > 0)
                {
                    // Title
                    // 称号ID
                    var titleId = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                    // object
                    var title = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"titleId", titleId}};
                    // add
                    list.Add(title);
                }
                return new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"list", list}};
            }
            default:throw new System.ArgumentException(System.String.Format("unknown protocol define: {0}", protocol));
        }
    }
}